analyzers/src/SonarAnalyzer.CSharp/Extensions/TypeDeclarationSyntaxExtensions.cs

```diff
@@ -33,14 +33,19 @@ public static class TypeDeclarationSyntaxExtensions
         /// Returns a union of all the methods and local functions from a given type declaration.
         /// </summary>
         public static IEnumerable<IMethodDeclaration> GetMethodDeclarations(this TypeDeclarationSyntax typeDeclaration) =>
-            typeDeclaration
-                .Members
-                .OfType<MethodDeclarationSyntax>()
-                .SelectMany(method => GetLocalFunctions(method).Union(new List<IMethodDeclaration> { MethodDeclarationFactory.Create(method) }));
+            typeDeclaration.Members
+                           .OfType<MethodDeclarationSyntax>()
+                           .SelectMany(method => GetLocalFunctions(method).Union(new List<IMethodDeclaration> { MethodDeclarationFactory.Create(method) }));
 
-        private static IEnumerable<IMethodDeclaration> GetLocalFunctions(MethodDeclarationSyntax methodDeclaration)
-            => methodDeclaration.DescendantNodes()
-                .Where(member => member.IsKind(SyntaxKindEx.LocalFunctionStatement))
-                .Select(member => MethodDeclarationFactory.Create(member));
+        public static IEnumerable<IMethodDeclaration> GetMethodDeclarations(this CompilationUnitSyntax compilationUnitSyntax) =>
+            compilationUnitSyntax.Members.OfType<GlobalStatementSyntax>()
+                                         .Select(x => x.ChildNodes().FirstOrDefault(y => y.IsKind(SyntaxKindEx.LocalFunctionStatement)))
```

Comment:
Let's check that it's immediate ancestor, then we will be sure that we won't be able to capture something else

Reply:
If I understand correctly you are referring to `x.ChildNodes()`? This should return only the immediate ancestors and in the case of the `GlobalStatement` it should be only one.
Is there some case I miss?

Reply:
Indeed!